diff -up filebench/procflow.c /tmp/filebench-1.4.9.1/procflow.c
--- filebench/procflow.c	2011-09-06 13:17:45.000000000 -0400
+++ /tmp/filebench-1.4.9.1/procflow.c	2011-09-24 00:23:34.298541435 -0400
@@ -514,7 +514,7 @@ static void
 procflow_sleep(procflow_t *procflow, int wait_cnt)
 {
 	while (procflow->pf_running & wait_cnt) {
-		(void) sleep(1);
+//		(void) sleep(1);
 		wait_cnt--;
 	}
 }
@@ -582,9 +582,13 @@ procflow_allstarted()
 	int running_procs = 0;
 	int ret = 0;
 
+//	filebench_log(LOG_INFO, "!!!Waiting on lock\n");
 	(void) ipc_mutex_lock(&filebench_shm->shm_procflow_lock);
 
-	(void) sleep(1);
+//	(void) sleep(1);
+//
+//
+//	filebench_log(LOG_INFO, "!!!Iterating\n");
 
 	while (procflow) {
 		int waits;
@@ -694,11 +703,14 @@ procflow_shutdown(void)
 
 			pid = procflow->pf_pid;
 #ifdef HAVE_SIGSEND
-			(void) sigsend(P_PID, pid, SIGUSR1);
+			(void) sigsend(P_PID, pid, SIGKILL);
 #else
-			(void) kill(pid, SIGUSR1);
+			(void) kill(pid, SIGKILL);
 #endif
-			procflow_wait(pid);
+			
+			filebench_log(LOG_INFO, "waiting for proclflow");
+//			procflow_wait(pid);
+			filebench_log(LOG_INFO, "waiting done");
 		}
 		(void) procflow_cleanup(procflow);
 		procflow = next_procflow;
diff -up filebench/threadflow.c /tmp/filebench-1.4.9.1/threadflow.c
--- filebench/threadflow.c	2011-09-06 13:17:45.000000000 -0400
+++ /tmp/filebench-1.4.9.1/threadflow.c	2011-09-24 00:23:34.328499963 -0400
@@ -329,9 +329,10 @@ threadflow_allstarted(pid_t pid, threadf
 		    threadflow->tf_name,
 		    threadflow->tf_instance);
 
-		waits = 10;
+		waits = 1000000;
 		while (waits && (threadflow->tf_running == 0) &&
 		    (filebench_shm->shm_f_abort == 0)) {
+//			filebench_log(LOG_INFO, "th_wait: %d\n", waits);
 			(void) ipc_mutex_unlock(
 			    &filebench_shm->shm_threadflow_lock);
 			if (waits < 3)
@@ -340,8 +341,9 @@ threadflow_allstarted(pid_t pid, threadf
 				    pid,
 				    threadflow->tf_name,
 				    threadflow->tf_instance);
-
-			(void) sleep(1);
+//			filebench_log(LOG_INFO, "3sleeping-before");
+//			(void) sleep(1);
+//			filebench_log(LOG_INFO, "3sleeping-after");
 			(void) ipc_mutex_lock(
 			    &filebench_shm->shm_threadflow_lock);
 			waits--;
